### ISSUE TITLE
Views

### ISSUE THREAD

**Issue by richardolsson** [zetkin/app.zetkin.org#451]
The "Views" feature is the main way of interacting with multiple people in the people database in Zetkin. Views let the user create spreadsheet-like tables of people with configurable columns useful when managing the database, organizing, analyzing data and more.

## Definitions of key terminology
* View: A sheet/table where every row is one person, and every column can contain some information related to that person
* Data set: A collection of people being displayed as rows in a view

## Target audience
These are the roles that will most likely use this feature:

* Database managers/administrators, i.e. people whose main job is to update person records, clean data, etc
* Organizers, i.e. people who are using Views as a means to support the social interactions of organizing, e.g. to find suitable people to reach out to, look up someone's contact information or to keep track of whom they've spoken to

## Requirements (job stories)
Below are job stories that describe the requirements for this feature.

### Need to have
- [x] When I have an organizing or reporting task that could be solved by assembling a spreadsheet, I want to be able to create a custom view with only the data that I need, so that I don't have to export data and collect it outside of Zetkin
- [x] When I'm creating a new view, I want to give it a name so that I can easily find it later
- [x] When I'm about to create a new view, I want to see all the saved views in the organization so that I can see if there is already a suitable one for my needs before I create one
- [x] When I'm browsing views, I want to see who created them so that I can talk to them if I have any questions
- [x] When I no longer need a view, I want to be able to delete it so that the list does not become cluttered
- [x] When I'm building my view or working with it later, I want to be able to add columns so that I can display and work with only the data that I need for the task at hand
- [x] When I'm adding a column, I want to be able to name it so that the resulting view communicates the content and purpose of the column
- [x] When I'm adding a column, I want it to automatically populate with data from Zetkin so that everything stays in sync
- [x] When I'm adding a column, I want to be able to choose from one of several types so that I can work with any type of data from Zetkin in the view. These are the options currently:
- [x] When I'm choosing the type of column, I want to be able to choose
  - [x] …a field column that renders a field from a person record, so that I can see things like names, phone numbers and addresses in views
  - [x] …a Smart Search query cross-reference column and pick a Smart Search, so that I can see whether each person in the view also matches that Smart Search query
  - [x] …a survey response column and pick an existing survey and question, so that I can see for every person what they have responded (if anything) to that question
  - [x] …a survey submit date column and pick a survey, so that I can see for every person what date they submitted that survey (if at all)
  - [x] …a "notes" column, so that I can see the most recent note on each person
  - [x] …a tag column and pick a tag, so I can toggle that tag on/off for people in the view individually
  - [x] …a "toggle" column that I can toggle on/off for each person
  - [x] …a person column, so that I can assign and reassign an organizer to each person record in the view
- [x] When I don't need a column, I want to be able to delete it so that I can free upp space in the view and display only the data that I need for the task at hand
- [x] When I'm working in a view, I want to be able to add an existing person record as a line, so that the columns auto-populate for that new person
- [x] When I'm adding a person to the view and there is already a row for that person, I want to be alerted so that I can easily find the existing row without adding the same row twice
- [x] When I'm trying to add a record and cannot find a person, I want to be able to create that record in Zetkin so that I can add the person to my view without having to navigate away and create them separately
- [x] When I'm navigating my view, I want to be able to hide all unnecessary chrome, so that I can browse the view unincumbered and see as many rows and columns as possible in my browser window
- [x] When a person is no longer relevant to the work I'm doing in my view, I want to be able to remove them only from the view, so that they disappear from the view but still exist in Zetkin
- [x] When I'm trying to make sense of the data in my view, I want to be able to sort the view on any column so that I can easily see the most recent/most relevant records on top
- [x] When I'm looking for patterns in my view, I want to be able to filter the columns based on the data therein so that I can hide those rows that are not relevant to what I'm doing

### Nice to have / known future improvements
- [ ] When I have a lot of views, I want to be able to structure them in folders so that I can keep related views together and unrelated views apart
- [ ] When my organization has a lot of views, I want some of them to be just mine, so that I can work with them without interference from, or interfering others
- [ ] When I've created a view which may be relevant to others, I want to be able to share it even with non-officials, so that they can use the view in their organizing without getting full access to Zetkin
- [ ] When my needs change, I want to be able to reorder the columns in my view, so that the order feels natural for the task at hand
- [ ] When I'm creating a new column, I want to be able to create it anywhere in the view so that I don't have to move it after creating it for it to appear in the right place
- [ ] When I'm building my view, I want to be able to add many columns at once so that I don't have to repeat steps to get started with a view that should contain multiple fields (e.g. e-mail address, phone number, join date etc)
- [x] When I'm about to create a new view, I want to be able to use an existing one as a template or copy an existing one, so that I don't have to start from scratch
- [x] When I'm adding a column, I want to be able to not pick a name manually, but have it be selected based on type, so that by default it reflects the content that will be loaded in the column
- [ ] When I'm brainstorming people to add to my view as a shortlist, I want to be able to define a Smart Search query and see the resulting data set in my view, so that I can add the ones I want to the view from there
- [x] When I'm using a Smart Search query to populate my view, I want to be able to save it in the view so that every time I come back the view contains the people currently matching that query
- [ ] When I'm creating a shortlist of people, I want to be able to switch between different data sets in my view so that I can compare and find the ones to add

## Solution
Wireframes are here: https://whimsical.com/gen-3-wireframes-Pnmi9fMGBHriA2tQrfZ4xK

### Views listing
Below are some key changes compared to Gen2 (see wireframes for details):
* Lists all views in a sortable table
* Contains suggestions for your most recently created views

### Single-view page
Below are some key changes compared to Gen2 (see wireframes for details):

* An empty view allows the user to pick between populating it "manually" (searching for one person at a time) or dynamically using Smart Search
* Users can quickly jump between views using a new concept of drop downs next to the title of the view

## Tasks
- [x] #467
- [x] #486
- [x] #497
- [x] #487
- [x] #468
- [x] #469
- [x] #470
- [x] #471
- [x] #472
- [x] #473
- [x] #474
- [x] #475
- [x] #493
- [x] #476
- [ ] #477
- [x] #478
- [x] #479
- [x] #480
- [x] #481
- [x] #482
- [x] #483

**Comment by richardolsson** [comment]
> @richardolsson I've done a first pass at the wireframes for this: https://whimsical.com/gen-3-wireframes-Pnmi9fMGBHriA2tQrfZ4xK

Awesome! It's so much easier to discuss when there's something visual to talk around. My answers below, and then some thoughts.

> **Manually adding people** There are a few stories that relate to adding new rows (e.g. people) directly to a view. I guess the way that I was thinking of this is that you're generating different views of the database based on filters, rather than directly handpicking people to add to your list. Could you explain a bit further why users should be able to manually add people?

There are two main ways that views are used – either populated dynamically with rows based on a Smart Search query, or populated manually by the user. What you are describing is the first approach. There are some use cases for views that would not be covered by this, for example:

* You're a team of organizers sitting in a meeting, brainstorming people to reach out to as the first batch of activists to mobilize to a campaign (e.g. phone bankers that will then mobilize further). You all blurt out names of people you can think of, and someone puts them one by one into a view.
* You're trying to find people to talk to about a new group that you're starting, and you want to reach out to people who might be interested based on some survey. You create a view and populate it using Smart Search based on survey data. You skim through the list and "save" those that seem interesting into a shortlist. In the future, you mainly work from that shortlist.

These are important use cases, but I'm open to alternative solutions.

> My hesitation is that this will make it a bit too easy to add a person to the database, and it lends itself towards adding someone in a kind of messy and manual way, without all the right data. My assumption is that most organisations have some kind of joining process, where people usually join through a form and perhaps occasionally over the phone or in person. So I don't see the use case for adding them in this situation.

I see the problem. In Gen2 we didn't allow creating people from views, and there have been situations where that has been impeding. Sometimes you want to add a new person that you haven't organized before. Maybe one solution could be to clearly warn about creating new people, or by opening a "new person" modal to add some friction.

> **Smart Search**
> 
> The other story I was wondering about is this one: "…a Smart Search query cross-reference column and pick a Smart Search, so that I can see whether each person in the view also matches that Smart Search query"
> 
> In my wireframes, it's possible to pick an existing Smart Search from a dropdown. However, it could be quite useful to perform a Smart Search directly from the Configure panel, rather than only using an existing one or navigating away and making one, then coming back to the view. What do you think? Where do people actually create these Smart Searches anyway?

I completely agree. In Gen2 you have to create the Smart Search (which happens under Manage > Smart Search) and then use it in your view, e.g. to filter the view or to add a cross-reference column. It would be nice to not have to take that extra step.

The problem is that the API requires a Smart Search to exist before it can be used – you can't search and then discard it. We could work around this right now by allowing the user to create a Smart Search through the views UI, even though it's actually a standalone Smart Search query.

In a future version of the API my idea is to allow views to include a dedicated Smart Search query that can be edited in the view and is not used (or available) elsewhere.

**Comment by richardolsson** [comment]
Here are some other thoughts:

## "Create" actions
![image](https://user-images.githubusercontent.com/550212/137132173-9abd0a77-52ee-4521-982e-2f292b0d902f.png)

In other places in the Gen3 UI we have used the MUI [Speed Dial component](https://mui.com/components/speed-dial/) to trigger "create" flows, by opening modals where the user fills in the necessary information.

In Gen2 we tried something a little bit different for views, which was to create the view with an "untitled" title, and then let the user change they made it into the view. We did that to mimic the way that Google Drive/Docs work, since views is in many ways a replacement for Spreadsheets. I think it was nice, and the inconsistency is warranted since views are so different than other types of Zetkin data (e.g. campaigns). But I'm curious to hear your thoughts!

## Edit/delete actions
In other places in Gen3, we have used a "cogwheel menu" in the top right with options to delete or edit objects. It could use some visual polish, but I think the UX is sound and makes sense to use for views as well.

What do you think? Below are screenshots of what it looks like.

### Before opening
![image](https://user-images.githubusercontent.com/550212/137134467-16d2a48b-d006-4de0-9fc8-f865b174dc80.png)

### After opening
![image](https://user-images.githubusercontent.com/550212/137134558-5438b145-9c1b-45fb-bd8d-e0d8859199ed.png)



## Suggested views
![image](https://user-images.githubusercontent.com/550212/137133428-c250ebf5-2086-419b-b9b1-c80a82fe3855.png)

I think this is nice! What do you reckon we should base those suggestions on? Most recently edited? Only views owned by the user?

**Comment by GemCopeland** [comment]
> Here are some other thoughts:
> 
> ## "Create" actions
> ![image](https://user-images.githubusercontent.com/550212/137132173-9abd0a77-52ee-4521-982e-2f292b0d902f.png)
> 
> In other places in the Gen3 UI we have used the MUI [Speed Dial component](https://mui.com/components/speed-dial/) to trigger "create" flows, by opening modals where the user fills in the necessary information.
> 
> In Gen2 we tried something a little bit different for views, which was to create the view with an "untitled" title, and then let the user change they made it into the view. We did that to mimic the way that Google Drive/Docs work, since views is in many ways a replacement for Spreadsheets. I think it was nice, and the inconsistency is warranted since views are so different than other types of Zetkin data (e.g. campaigns). But I'm curious to hear your thoughts!
> 
> ## Edit/delete actions
> In other places in Gen3, we have used a "cogwheel menu" in the top right with options to delete or edit objects. It could use some visual polish, but I think the UX is sound and makes sense to use for views as well.
> 
> What do you think? Below are screenshots of what it looks like.
> 
> ### Before opening
> ![image](https://user-images.githubusercontent.com/550212/137134467-16d2a48b-d006-4de0-9fc8-f865b174dc80.png)
> 
> ### After opening
> ![image](https://user-images.githubusercontent.com/550212/137134558-5438b145-9c1b-45fb-bd8d-e0d8859199ed.png)

I forgot about the speed-dial! Good that you flagged this. I've been looking at the Gen2 version that I forgot about the work that's already done with Gen3.

Potentially instead of the cog, we should use the MUI vertical dots? But this is a small detail.

I'm going to make a general page template with these patterns as I think it's important they're consistent throughout. 

> ## Suggested views
> ![image](https://user-images.githubusercontent.com/550212/137133428-c250ebf5-2086-419b-b9b1-c80a82fe3855.png)
> 
> I think this is nice! What do you reckon we should base those suggestions on? Most recently edited? Only views owned by the user?

I think recently edited or opened by the user.

**Comment by richardolsson** [comment]
I also left a comment in Whimsical re Smart Search interface, but I'll repeat it here.

We have developed a Smart Search UX in Gen3 that I think is quite nice, originally tracked as #336. You can try it out by clicking "Create Smart Search" here: http://app.dev.zetkin.org/organize/1/campaigns/2/calendar/tasks/43/assignees

I think we should use this UI for when we're creating Smart Search queries in views as well. And then maybe, if you have feedback, we can make changes to it consistently across Gen3 in a future release.

**Comment by GemCopeland** [comment]
Thanks for your detailed answer… it feels like we're getting there, almost.

I mapped out those scenarios you mentioned:
![image](https://user-images.githubusercontent.com/2996847/138325646-a6b1c8a4-824a-4b48-92ef-c61a82f1bd61.png)
[Full version here](https://whimsical.com/views-FJGzLB4gEsBWy9yfsdUAWE)

And I then made [wireframes](https://whimsical.com/gen-3-wireframes-Pnmi9fMGBHriA2tQrfZ4xK) for each of the steps in each scenario.

I thought that rather than introducing a new concept ("static or dynamic"), could we just reuse the Smart Search for creating views? In a way, it's the same thing: start with an empty list or a full list, apply some parameters. So it made sense to me to just repeat that interface, although I feel like there will be a technical reason that will make this difficult perhaps.

It's not as immediate as creating a new spreadsheet in Google Sheets, but I think that makes sense because it's much more powerful and complex than an empty spreadsheet.

Let me know what you think!
